### PR TITLE
Simplify debugging through synchronous data flow

### PR DESCRIPTION
**Pull request by hmoog**
# Description of change

To speed up the debugging process, we temporarily make the processing of blocks synchronous.

After stabilizing the code base and removing the last bugs, we can re-introduce multi-threading on a component by component basis without having to look for race conditions in all components simultaneously.

In addition to the named changes, this PR already fixes a few deadlocks in the data flow.

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/network/p2p/manager.go [hunk 1]
```diff
@@ -85,12 +86,12 @@ func NewManager(libp2pHost host.Host, local *peer.Local, log *logger.Logger) *Ma
 // Stop stops the manager and closes all established connections.
 func (m *Manager) Stop() {
 	m.stopMutex.Lock()
-	defer m.stopMutex.Unlock()
-
 	if m.isStopped {
 		return
 	}
 	m.isStopped = true
+	m.stopMutex.Unlock()
```
Comment (piotrm50) [inline packages/network/p2p/manager.go]: stop mutex is not unlocked in the `if` above

File: packages/protocol/engine/ledger/mempool/realitiesledger/booker.go [hunk 2]
```diff
@@ -180,18 +187,27 @@ func (b *booker) forkTransaction(ctx context.Context, txID utxo.TransactionID, o
...
+			TransactionID:      consumingTxMetadata.ID(),
+			AddedConflictID:    forkedConflictID,
+			RemovedConflictIDs: previousParentConflicts,
+			Context:            ctx,
+		})
 
 		walker.PushAll(consumingTxMetadata.OutputIDs().Slice()...)
 	})
+
+	//
```
Comment (piotrm50) [inline packages/protocol/engine/ledger/mempool/realitiesledger/booker.go]:
```suggestion
```

File: packages/protocol/engine/tangle/blockdag/inmemoryblockdag/blockdag.go [hunk 3]
```diff
@@ -139,8 +141,8 @@ func (b *BlockDAG) Attach(data *models.Block) (block *blockdag.Block, wasAttache
 	if block, wasAttached, err = b.attach(data); wasAttached {
 		b.events.BlockAttached.Trigger(block)
 
-		b.solidifierMutex.RLock()
-		defer b.solidifierMutex.RUnlock()
+		//b.solidifierMutex.Lock()
+		//defer b.solidifierMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/blockdag/inmemoryblockdag/blockdag.go]:
```suggestion
		//b.solidifierMutex.RLock()
		//defer b.solidifierMutex.RUnlock()
```

File: packages/protocol/engine/tangle/blockdag/inmemoryblockdag/blockdag.go [hunk 4]
```diff
@@ -200,10 +202,10 @@ func (b *BlockDAG) SetOrphaned(block *blockdag.Block, orphaned bool) (updated bo
 }
 
 func (b *BlockDAG) PromoteFutureBlocksUntil(index slot.Index) {
-	b.solidifierMutex.RLock()
-	defer b.solidifierMutex.RUnlock()
-	b.futureBlocksMutex.Lock()
-	defer b.futureBlocksMutex.Unlock()
+	//b.solidifierMutex.Lock()
+	//defer b.solidifierMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/blockdag/inmemoryblockdag/blockdag.go]:
```suggestion
	//b.solidifierMutex.RLock()
	//defer b.solidifierMutex.RUnlock()
```

File: packages/protocol/engine/tangle/blockdag/testframework.go [hunk 5]
```diff
@@ -81,10 +82,10 @@ func (t *TestFramework) SlotTimeProvider() *slot.TimeProvider {
 func (t *TestFramework) IssueBlocks(blockAliases ...string) *TestFramework {
 	for _, alias := range blockAliases {
 		currentBlock := t.ModelsTestFramework.Block(alias)
-
-		t.workerPool.Submit(func() {
-			_, _, _ = t.Instance.Attach(currentBlock)
-		})
+		fmt.Println("issue", alias)
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/blockdag/testframework.go]:
```suggestion

```

File: packages/protocol/engine/tangle/booker/markerbooker/booker.go [hunk 6]
```diff
@@ -318,8 +311,8 @@ func (b *Booker) BlockFloor(marker markers.Marker) (floorMarker markers.Marker,
 
 // MarkerVotersTotalWeight retrieves Validators supporting a given marker.
 func (b *Booker) MarkerVotersTotalWeight(marker markers.Marker) (totalWeight int64) {
-	b.sequenceEvictionMutex.RLock()
-	defer b.sequenceEvictionMutex.RUnlock()
+	//b.sequenceEvictionMutex.Lock()
+	//defer b.sequenceEvictionMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/booker/markerbooker/booker.go]:
```suggestion
	//b.sequenceEvictionMutex.RLock()
	//defer b.sequenceEvictionMutex.RUnlock()
```

File: packages/protocol/engine/tangle/booker/markerbooker/booker.go [hunk 7]
```diff
@@ -334,8 +327,8 @@ func (b *Booker) MarkerVotersTotalWeight(marker markers.Marker) (totalWeight int
 
 // SlotVotersTotalWeight retrieves the total weight of the Validators voting for a given slot.
 func (b *Booker) SlotVotersTotalWeight(slotIndex slot.Index) (totalWeight int64) {
-	b.sequenceEvictionMutex.RLock()
-	defer b.sequenceEvictionMutex.RUnlock()
+	//b.sequenceEvictionMutex.Lock()
+	//defer b.sequenceEvictionMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/booker/markerbooker/booker.go]:
```suggestion
	//b.sequenceEvictionMutex.RLock()
	//defer b.sequenceEvictionMutex.RUnlock()
```

File: packages/protocol/engine/tangle/booker/markerbooker/booker.go [hunk 8]
```diff
@@ -364,9 +357,25 @@ func (b *Booker) GetAllAttachments(txID utxo.TransactionID) (attachments *advanc
 	return b.attachments.GetAttachmentBlocks(txID)
 }
 
+// storeNewBlock tries to store a new Block in the in-memory storage of the Booker and returns if the Block was stored.
+func (b *Booker) storeNewBlock(block *booker.Block) bool {
+	b.evictionMutex.Lock()
+	defer b.evictionMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/booker/markerbooker/booker.go]: This used to be a RLock

```suggestion
	b.evictionMutex.RLock()
	defer b.evictionMutex.RUnlock()
```

File: packages/protocol/engine/tangle/booker/markerbooker/booker.go [hunk 9]
```diff
@@ -364,9 +357,25 @@ func (b *Booker) GetAllAttachments(txID utxo.TransactionID) (attachments *advanc
...
+	}
+
+	return true
+}
+
 func (b *Booker) ProcessForkedMarker(marker markers.Marker, forkedConflictID utxo.TransactionID, parentConflictIDs utxo.TransactionIDs) {
-	b.sequenceEvictionMutex.RLock()
-	defer b.sequenceEvictionMutex.RUnlock()
+	b.sequenceEvictionMutex.Lock()
+	defer b.sequenceEvictionMutex.Unlock()
```
Comment (piotrm50) [inline packages/protocol/engine/tangle/booker/markerbooker/booker.go]:
```suggestion
	b.sequenceEvictionMutex.RLock()
	defer b.sequenceEvictionMutex.RUnlock()
```

**Review by piotrm50 — CHANGES_REQUESTED** [review]

File: packages/network/p2p/manager.go [hunk 10]
```diff
@@ -233,10 +249,11 @@ func (m *Manager) addNeighbor(ctx context.Context, p *peer.Peer, group Neighbors
 		return errors.WithStack(ErrLoopbackNeighbor)
 	}
 	m.stopMutex.RLock()
-	defer m.stopMutex.RUnlock()
 	if m.isStopped {
 		return ErrNotRunning
```
Comment (hmoog) [inline packages/network/p2p/manager.go:250]: Unlock is missing

**Review by piotrm50 — APPROVED** [review]

**Pull request merged**